src/Groups/GAPGroups.jl

```diff
@@ -1278,25 +1280,119 @@ false
+power. In particular, the trivial group is a ``p``-group for every prime.
+
+# Examples
+```jldoctest
+julia> is_pgroup(symmetric_group(1))
+true
+
+julia> is_pgroup(symmetric_group(2))
+true
+
+julia> is_pgroup(symmetric_group(3))
+false
+
+```
+"""
+@gapattribute is_pgroup(G::GAPGroup) = GAP.Globals.IsPGroup(G.X)::Bool
+
+
+"""
+    is_pgroup_with_prime(G)
+
 Return `(true, nothing)` if `G` is the trivial group,
 `(true, p)` if the order of every element in `G` is a power of a prime `p`,
 and `(false, nothing)` otherwise.
 
 For finite groups `G`, the first return value is `true` if and only if
 the order of `G` is a prime power.
+
+# Examples
+```jldoctest
+julia> is_pgroup_with_prime(symmetric_group(1))
+(true, nothing)
+
+julia> is_pgroup_with_prime(symmetric_group(2))
+(true, 2)
+
+julia> is_pgroup_with_prime(symmetric_group(3))
+(false, nothing)
+
+```
 """
-function is_pgroup(G::GAPGroup)
-   if GAPWrap.IsPGroup(G.X)
-      p = GAP.Globals.PrimePGroup(G.X)
-      if p != GAP.Globals.fail
-         return true, fmpz(p)  # TODO: allow specifying the type used for the prime
-      else
-         return true, nothing
-      end
-   end
-   return false, nothing
+function is_pgroup_with_prime(G::GAPGroup)
+  is_trivial(G) && return true, nothing
+  if is_pgroup(G)
+     p = GAP.Globals.PrimePGroup(G.X)
+     if p != GAP.Globals.fail
+        return true, fmpz(p)  # TODO: allow specifying the type used for the prime
+     else
+        return true, nothing
+     end
+  end
+  return false, nothing
 end
 
+"""
+    prime_pgroup(::Type{T} = fmpz, G::GAPGroup) where T <: IntegerUnion
+
+Return the prime ``p`` if `G` is a non-trivial ``p``-group.
+
+An exception is thrown if `G` is not a ``p``-group or
+is a trivial group.
+
+# Examples
+```jldoctest
+julia> prime_pgroup(quaternion_group(8))
+2
+
+julia> prime_pgroup(UInt16, quaternion_group(8))
+0x0002
+
+julia> prime_pgroup(symmetric_group(1))
+ERROR: not supported for trivial groups
+
+julia> prime_pgroup(symmetric_group(3))
+ERROR: only supported for non-trivial p-groups
+
+```
+"""
+function prime_pgroup(::Type{T}, G::GAPGroup) where T <: IntegerUnion
+  return T(_prime_pgroup(G))
+end
+
+prime_pgroup(G::GAPGroup) = prime_pgroup(fmpz, G)
+
+@gapattribute function _prime_pgroup(G::GAPGroup)
+  is_trivial(G) && error("not supported for trivial groups")
+  is_pgroup(G) || error("only supported for non-trivial p-groups")
+  p = GAP.Globals.PrimePGroup(G.X)
+  p === GAP.Globals.fail && error("not available")
+  return p::GapInt
+end
+
+"""
+    has_prime_pgroup(G::GAPGroup)
+
+Return `true` if the value for `prime_pgroup(G)` has already been computed.
+"""
+has_prime_pgroup(G::GAPGroup) = has__prime_pgroup(G)
+
+"""
+    set_prime_pgroup(G::GAPGroup, p::IntegerUnion)
+
+Set the value for `prime_pgroup(G)` to `p` if it has't been set already.
+"""
+function set_prime_pgroup(G::GAPGroup, p::IntegerUnion)
+  set__prime_pgroup(G, GAP.julia_to_gap(p))
+end
```

Comment:
As far as I understand, the complicated construction with `_prime_pgroup` and `prime_pgroup`, and explicitly defined `has_prime_pgroup` and `set_prime_pgroup(` is needed only in order to avoid duplicating the code of the one-argument method (or to avoid converting the GAP result first to `fmpz` and then to the intended type, at runtime). I do not see a really nice solution.

Reply:
Yeah, that's the reason. Perhaps a nicer solution (e.g. via an additional macro?) might be possible, but I think it's better to do this a few more times to see if there is a clear pattern.

Reply:
I've added a comment explaining what is going on